tests/integration/relation_tests/new_relations/test_charm_relations.py

```diff
@@ -94,6 +96,104 @@ async def test_database_relation_with_charm_libraries(ops_test: OpsTest):
+            unit.public_address in endpoints_str
+        ), f"unit {unit.name} not present in connection URI"
+
+    assert len(endpoints_str.split(",")) == len(
+        ops_test.model.applications[DATABASE_APP_NAME].units
+    ), "number of endpoints in replicaset URI do not match number of units"
+
+    await ops_test.model.applications[DATABASE_APP_NAME].add_units(count=2)
+    await ops_test.model.wait_for_idle(
+        apps=[DATABASE_APP_NAME], status="active", timeout=1000, wait_for_exact_units=4
+    )
+
+    endpoints_str = await get_application_relation_data(
+        ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME, "endpoints"
+    )
+    for unit in ops_test.model.applications[DATABASE_APP_NAME].units:
+        assert (
+            unit.public_address in endpoints_str
+        ), f"unit {unit.name} not present in connection URI after adding units"
+
+    assert len(endpoints_str.split(",")) == len(
+        ops_test.model.applications[DATABASE_APP_NAME].units
+    ), "number of endpoints in replicaset URI do not match number of units after adding units"
+
+    await ops_test.model.applications[DATABASE_APP_NAME].destroy_units(
+        f"{DATABASE_APP_NAME}/0", f"{DATABASE_APP_NAME}/1"
+    )
+    await ops_test.model.wait_for_idle(
+        apps=[DATABASE_APP_NAME], status="active", timeout=1000, wait_for_exact_units=2
+    )
+
+    endpoints_str = await get_application_relation_data(
+        ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME, "endpoints"
+    )
+    for unit in ops_test.model.applications[DATABASE_APP_NAME].units:
+        assert (
+            unit.public_address in endpoints_str
+        ), f"unit {unit.name} not present in connection URI after destroying units"
+
+    assert len(endpoints_str.split(",")) == len(
+        ops_test.model.applications[DATABASE_APP_NAME].units
+    ), "number of endpoints in replicaset URI do not match number of units after destroying units"
+
+    # check that the replica set with the remaining units has a primary
+    ip_addresses = endpoints_str.split(",")
+    try:
+        primary = await replica_set_primary(ip_addresses, ops_test)
+    except RetryError:
+        primary = None
+
+    # verify that the primary is not None
+    assert primary is not None, "replica set has no primary"
+
+    # test crud operations
+    connection_string = await get_application_relation_data(
+        ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME, "uris"
+    )
+    database = await get_application_relation_data(
+        ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME, "database"
+    )
+    client = MongoClient(
+        connection_string,
+        directConnection=False,
+        connect=False,
+        serverSelectionTimeoutMS=1000,
+        connectTimeoutMS=2000,
+    )
+
+    db = client[database]
+    test_collection = db["test_app_collection"]
+    ubuntu = {"release_name": "Focal Fossa", "version": 20.04, "LTS": True}
```

Comment:
seems like the data below is already written from the previous test. To avoid any false positives from the previous test, it would be best to clear the DB
```suggestion
    # clear collection  writes from previous test
    test_collection = db["test_app_collection"]
    test_collection.drop()
    test_collection = db["test_app_collection"]
    
    ubuntu = {"release_name": "Focal Fossa", "version": 20.04, "LTS": True}
```

Reply:
Done